src/chainbin/nnet3-chain-get-egs.cc

```diff
@@ -38,12 +38,49 @@ namespace nnet3 {
    you should do it later with nnet3-chain-normalize-egs.
 */
 
+static void GetIvectorsForEgs(int32 t, int32 left_context,
+                              int32 right_context, int32 frames_per_eg,
+                              int32 ivector_period,
+                              const MatrixBase<BaseFloat> *ivector_feats,
+                              NnetChainExample *eg) {
+  // try to get a representative iVector every ivector_period frames
+  // in an eg.
+  KALDI_ASSERT(ivector_feats->NumRows() > 0);
+  // num_ivectors_in_left_context is the num of ivectors for frames whose t < 0.
+  // It is used to compute the initial value of closest_frame
+  // num_ivectors is the num of ivectors for frames whose t >= 0.
+  // Both of them are computed according to how Round descriptor works, which
+  // basically returns floor(t / <t-modulus>) * <t-modulus>.
+  int32 num_ivectors_in_left_context =
+      std::ceil(1.0 * left_context / ivector_period); 
+  int32 num_ivectors = num_ivectors_in_left_context +
+      (frames_per_eg + right_context - 1) / ivector_period + 1;
+  Matrix<BaseFloat> ivectors(num_ivectors, ivector_feats->NumCols());
+  // closest_frame is the frame at whole multiples of ivector_period
+  int32 closest_frame = t - num_ivectors_in_left_context * ivector_period;
+  for (int32 n = 0; n < num_ivectors; n++) {
```

Comment:
It's helpful to think in a more "functional" way.  Given the desired frame, what is the closest frame that we have to it?
You are carrying too much state from one iteration to the next.